.circleci/ci-config.yml

```diff
@@ -0,0 +1,934 @@
+    docker: *docker
+    steps:
+      - checkout
+      - *skip_forked_pr
+      - *build
+      - run:
+          name: "Delete stage datasets"
+          command: |
+            export GOOGLE_APPLICATION_CREDENTIALS="/tmp/gcp.json"
+            echo "$GCLOUD_SERVICE_KEY" > "$GOOGLE_APPLICATION_CREDENTIALS"
+
+            PATH="venv/bin:$PATH" script/bqetl stage clean --dataset-suffix=$CIRCLE_SHA1 --delete-expired
+  manual-trigger-required-for-fork:
+    docker: *docker
+    steps:
+      - &skip_upstream
+        run:
+          name: Early return if this build is running on upstream
+          command: |
+            if [ -n "$CIRCLE_PR_NUMBER" ]; then
+              echo "Build on fork"
+            else
+              echo "Build on upstream"
+              circleci-agent step halt
+            fi
+      - checkout
+      - run:
+          name: Manually trigger integration tests for fork
+          # yamllint disable rule:line-length
+          command: |
+            apt update
+            apt install jq -y
+
+            CIRCLE_PR_BRANCH=`curl -s https://api.github.com/repos/${CIRCLE_PROJECT_USERNAME}/${CIRCLE_PROJECT_REPONAME}/pulls/${CIRCLE_PR_NUMBER} | jq -r '.head.label'`
+
+            echo "Integration tests for this fork need to be triggered manually"
+            echo "Users with write access to the repository can trigger" \
+              "integration tests by following these steps: "
+            echo "  Open the following page:"
+            echo "    https://github.com/mozilla/bigquery-etl/actions/workflows/push-to-upstream.yml"
+            echo "  Choose the 'Run workflow' dropdown and provide '$CIRCLE_PR_BRANCH' as parameter."
+
+            exit 1
+          # yamllint enable rule:line-length
+  deploy-to-pypi:
+    docker: *docker
+    steps:
+      - checkout
+      - run:
+          name: Install deployment tools
+          command: |
+            pip install --upgrade setuptools wheel twine
+      - run:
+          name: Create the distribution files
+          command: |
+            python -m build --sdist --wheel
+      - run:
+          name: Upload to PyPI
+          command: |
+            # Relies on the TWINE_USERNAME and TWINE_PASSWORD environment variables
+            # For more on twine, see:
+            #   https://twine.readthedocs.io/en/latest/
+            twine upload --skip-existing dist/*
+
+workflows:
+  version: 2
+  build:
+    jobs: &build_jobs
+      - manual-trigger-required-for-fork
+      - build
+      - verify-requirements
+      - verify-format-yaml
+  validate-sql:
+    when: 
+      or: 
+        - << pipeline.parameters.validate-sql >>
+    jobs:
+      - verify-format-sql
+      - generate-sql
+      - deploy-changes-to-stage:
+          requires:
+            - generate-sql
+      - dry-run-sql:
+          requires:
+            - deploy-changes-to-stage
+      - test-sql:
+          context: data-eng-circleci-tests
+          requires:
+            - deploy-changes-to-stage
```

Comment:
IIRC these take the majority of CI time. A short version of the proposal I was drafting:

It should be possible to run the sql jobs only on the affected artifacts. For e.g. if I change `table_a` which has upstream or downstream `table_b`, `udf_1`, `sql_generator_1`, and `view_a`, then only run the necessary generators and stage the necessary artifacts similar to the way `deploy-to-stage` works now. This would require some changes to sql generators since sometimes the dependencies aren't known until the generator gets run